File: backend/djadmin/monitor/log_health.py
"""日志采集链路对账：逐层比对「期望状态」与「集群/主机实际状态」。

存在意义是消除静默失败——索引模板没下发、pipeline 没发布、主机配置改了没同步，
这些问题原先只能等聚合报 400 或登上主机看 journalctl 才发现。
本模块全程只读，不做任何写入或自动修复。
"""
import json

from django.utils import timezone

from .fluent_bit import build_host_fragments
from .log_management import (
    build_index_template_body,
    build_index_template_name,
    build_ism_policy_body,
    build_ism_policy_name,
)
from .models import LogCollectionTarget, LogProcessingRule, LogRetentionTier
from .opensearch_client import OpenSearchClient, OpenSearchError

STATUS_OK = 'ok'
STATUS_WARN = 'warn'
STATUS_DRIFT = 'drift'
STATUS_ERROR = 'error'

# 数值越大越严重：明细项汇总成层状态、层状态汇总成整体状态都取最严重的一个。
_STATUS_RANK = {STATUS_OK: 0, STATUS_WARN: 1, STATUS_DRIFT: 2, STATUS_ERROR: 3}

DATA_FLOW_WINDOW_MINUTES = 15


def _worst(statuses):
    return max(statuses, key=lambda status: _STATUS_RANK.get(status, 0), default=STATUS_OK)


def _item(name, status, detail=''):
    return {'name': name, 'status': status, 'detail': detail}


def _layer(key, name, status, summary, items=None):
    return {'key': key, 'name': name, 'status': status, 'summary': summary, 'items': items or []}


def _layer_from_items(key, name, items, empty_summary):
    if not items:
        return _layer(key, name, STATUS_WARN, empty_summary)
    abnormal = [entry for entry in items if entry['status'] != STATUS_OK]
    summary = f'{len(items)} 项全部一致' if not abnormal else f'{len(abnormal)}/{len(items)} 项需要处理'
    return _layer(key, name, _worst([entry['status'] for entry in items]), summary, items)


def _is_not_found(exc):
    """OpenSearchError 消息格式为 '<HTTP 状态码>: <响应体>'，据此区分「对象不存在」与「集群异常」。"""
    return str(exc).startswith('404')


def _host_label(target):
    return str(getattr(target.host, 'ip', '') or '') or f'host-{target.host_id}'


def _check_index_template(client, cluster):
    """模板决定字段类型：dynamic 漏成 true 会把 keyword 建成 text，聚合类功能全部失效。"""
    prefix = cluster.index_prefix or 'logs'
    name = build_index_template_name(prefix)
    desired = build_index_template_body(prefix)['template']['mappings']

    try:
        response = client.get_index_template(name)
    except OpenSearchError as exc:
        if _is_not_found(exc):
            return _layer('index_template', '索引模板', STATUS_DRIFT, f'模板 {name} 不存在，新建索引会走动态映射')
        return _layer('index_template', '索引模板', STATUS_ERROR, f'读取模板失败: {exc}')

    templates = response.get('index_templates') or []
    actual = {}
    if templates:
        actual = ((templates[0].get('index_template') or {}).get('template') or {}).get('mappings') or {}

    desired_dynamic = desired.get('dynamic', True)
    actual_dynamic = actual.get('dynamic', True)
    items = [_item(
        'dynamic',
        STATUS_OK if actual_dynamic == desired_dynamic else STATUS_DRIFT,
        f'期望 {desired_dynamic}，实际 {actual_dynamic}',
    )]

    actual_properties = actual.get('properties') or {}
    for field, spec in (desired.get('properties') or {}).items():
        current = actual_properties.get(field)
        if not current:
            items.append(_item(field, STATUS_DRIFT, f"缺失，期望 {spec.get('type')}"))
        elif current.get('type') != spec.get('type'):
            items.append(_item(field, STATUS_DRIFT, f"类型不符：期望 {spec.get('type')}，实际 {current.get('type')}"))
        else:
            items.append(_item(field, STATUS_OK, str(spec.get('type'))))

    return _layer_from_items('index_template', '索引模板', items, '模板无字段定义')


def _ism_signature(policy):
    """只提取影响保留行为的字段：集群会补 policy_id/last_updated_time 等元数据，全量比对必然误报。"""
    states = {str(state.get('name')): state for state in (policy.get('states') or [])}
    hot = states.get('hot') or {}
    rollover = next(
        (action['rollover'] for action in (hot.get('actions') or []) if isinstance(action, dict) and 'rollover' in action),
        {},
    )
    delete_after = next(
        (
            (transition.get('conditions') or {}).get('min_index_age')
            for transition in (hot.get('transitions') or []) if transition.get('state_name') == 'delete'
        ),
        None,
    )
    templates = policy.get('ism_template') or []
    if isinstance(templates, dict):
        templates = [templates]
    return {
        'rollover_size': rollover.get('min_primary_shard_size'),
        'rollover_age': rollover.get('min_index_age'),
        'delete_after': delete_after,
        'index_patterns': sorted(
            str(pattern)
            for entry in templates
            for pattern in (entry.get('index_patterns') or [])
        ),
    }


def _check_ism_policies(client, cluster):
    """ISM 缺失时索引既不滚动也不清理，磁盘会被慢慢撑满，属于静默故障。"""
    prefix = cluster.index_prefix or 'logs'
    items = []
    for tier in LogRetentionTier.objects.filter(enabled=True):
        name = build_ism_policy_name(prefix, tier.code)
        desired = _ism_signature(build_ism_policy_body(prefix, tier)['policy'])
        try:
            actual = _ism_signature((client.get_ism_policy(name) or {}).get('policy') or {})
        except OpenSearchError as exc:
            if _is_not_found(exc):
                items.append(_item(name, STATUS_DRIFT, '策略不存在，索引不会自动滚动与清理'))
            else:
                items.append(_item(name, STATUS_ERROR, str(exc)[:200]))
            continue
        if actual == desired:
            items.append(_item(name, STATUS_OK, f'保留 {tier.retention_days} 天'))
        else:
            differing = [key for key, value in desired.items() if actual.get(key) != value]
            items.append(_item(name, STATUS_DRIFT, f"与档位配置不一致: {', '.join(differing)}"))

    return _layer_from_items('ism_policies', '保留策略', items, '没有启用中的保留档位')


def _pipeline_signature(body):
    """processors 有序，直接按顺序序列化比对；描述等非行为字段不参与。"""
    return json.dumps(
        {'processors': (body or {}).get('processors') or [], 'on_failure': (body or {}).get('on_failure') or []},
        sort_keys=True,
        ensure_ascii=False,
    )


def _check_pipelines(client, cluster):
    """pipeline 没发布时日志照样写入，只是不被解析，界面上看不出任何异常。"""
    items = []
    for rule in LogProcessingRule.objects.filter(cluster=cluster):
        try:
            remote = client.get_pipeline(rule.name) or {}
        except OpenSearchError as exc:
            if _is_not_found(exc):
                items.append(_item(rule.name, STATUS_DRIFT, '集群上不存在该 pipeline，日志不会被解析'))
            else:
                items.append(_item(rule.name, STATUS_ERROR, str(exc)[:200]))
            continue

        body = remote.get(rule.name) or {}
        if not body:
            items.append(_item(rule.name, STATUS_DRIFT, '集群上不存在该 pipeline，日志不会被解析'))
        elif _pipeline_signature(body) != _pipeline_signature(rule.pipeline_body):
            items.append(_item(rule.name, STATUS_DRIFT, '集群上的 pipeline 与页面配置不一致，需重新发布'))
        else:
            processors = (rule.pipeline_body or {}).get('processors') or []
            items.append(_item(rule.name, STATUS_OK, f'{len(processors)} 个处理器'))

    return _layer_from_items('pipelines', '解析规则', items, '尚未配置解析规则')


def _check_host_configs(cluster):
    """比对「当前应有的配置指纹」与「最近一次下发记录的指纹」，暴露改了配置但没下发的主机。

    注意这里比的是数据库记录，不回读主机文件；主机被人手工改过仍会显示一致。
    """
    items = []
    targets = LogCollectionTarget.objects.filter(managed_enabled=True).select_related('host')
    for target in targets:
        label = _host_label(target)
        if not target.agent_installed:
            items.append(_item(label, STATUS_WARN, 'Fluent Bit 未安装'))
            continue
        try:
            fragments = build_host_fragments(target.host, cluster)
        except ValueError as exc:
            items.append(_item(label, STATUS_ERROR, f'配置渲染失败: {exc}'))
            continue

        if not fragments['inputs']:
            # 主机上没有任何逻辑服务日志要采，本就无需下发，报漂移属于噪音。
            items.append(_item(label, STATUS_OK, '无采集项'))
        elif not target.config_fingerprint:
            items.append(_item(label, STATUS_DRIFT, '从未下发过采集配置'))
        elif fragments['fingerprint'] != target.config_fingerprint:
            items.append(_item(label, STATUS_DRIFT, '配置已变更但未下发'))
        else:
            items.append(_item(label, STATUS_OK, f"{len(fragments['inputs'])} 个采集项"))

    return _layer_from_items('host_configs', '主机配置', items, '没有纳管中的采集目标')


def _check_runtime():
    """运行状态取自数据库缓存，反映最近一次探测结果，不是实时探活。"""
    items = []
    targets = LogCollectionTarget.objects.filter(
        managed_enabled=True, agent_installed=True,
    ).select_related('host')
    for target in targets:
        label = _host_label(target)
        if target.runtime_status == LogCollectionTarget.RuntimeStatus.RUNNING:
            items.append(_item(label, STATUS_OK, '运行中'))
        elif target.runtime_status == LogCollectionTarget.RuntimeStatus.ERROR:
            items.append(_item(label, STATUS_ERROR, (target.last_error or '异常')[:200]))
        elif target.runtime_status == LogCollectionTarget.RuntimeStatus.STOPPED:
            items.append(_item(label, STATUS_DRIFT, '已停止'))
        else:
            items.append(_item(label, STATUS_WARN, '状态未知，需刷新'))

    return _layer_from_items('runtime', '采集进程', items, '没有已安装 Fluent Bit 的主机')


def _check_data_flow(client, cluster):
    """前面几层全绿也可能没数据，这一层是唯一能证明链路真正通了的证据。"""
    prefix = cluster.index_prefix or 'logs'
    body = {
        'size': 0,
        'query': {'range': {'@timestamp': {'gte': f'now-{DATA_FLOW_WINDOW_MINUTES}m'}}},
        'aggs': {'by_service': {'terms': {'field': 'service', 'size': 50}}},
    }
    try:
        result = client.search(f'{prefix}-*', body)
    except OpenSearchError as exc:
        return _layer('data_flow', '数据写入', STATUS_ERROR, f'查询失败: {exc}')

    total = ((result.get('hits') or {}).get('total') or {}).get('value') or 0
    buckets = ((result.get('aggregations') or {}).get('by_service') or {}).get('buckets') or []
    items = [_item(str(bucket['key']), STATUS_OK, f"{bucket['doc_count']} 条") for bucket in buckets]

    if not total:
        return _layer(
            'data_flow', '数据写入', STATUS_WARN,
            f'最近 {DATA_FLOW_WINDOW_MINUTES} 分钟没有新日志写入', items,
        )
    return _layer(
        'data_flow', '数据写入', STATUS_OK,
        f'最近 {DATA_FLOW_WINDOW_MINUTES} 分钟写入 {total} 条，覆盖 {len(buckets)} 个服务', items,
    )


def collect_log_pipeline_health(cluster):
    """按「存储 → 解析 → 采集 → 数据」的顺序逐层对账，任一层异常都会冒泡到整体状态。"""
    client = OpenSearchClient(cluster)
    layers = [
        _check_index_template(client, cluster),
        _check_ism_policies(client, cluster),
        _check_pipelines(client, cluster),
        _check_host_configs(cluster),
        _check_runtime(),
        _check_data_flow(client, cluster),
    ]
    return {
        'status': _worst([layer['status'] for layer in layers]),
        'checked_at': timezone.now().isoformat(),
        'layers': layers,
    }

File: autoadmin/internal/monitor/log_health.go
package monitor

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	"autoadmin/internal/api/response"

	"github.com/gin-gonic/gin"
)

// 日志采集链路对账：逐层比对「期望状态」与「集群/主机实际状态」，对应 Django monitor/log_health.py。
// 存在意义是消除静默失败——模板没下发、policy 没发布、配置改了没同步，这些问题
// 原先只能等聚合报错或登上主机看 journalctl 才发现。全程只读，不做任何写入或自动修复。

const (
	logHealthOK    = "ok"
	logHealthWarn  = "warn"
	logHealthDrift = "drift"
	logHealthError = "error"

	logHealthDataFlowWindowMinutes = 15
)

var logHealthStatusRank = map[string]int{logHealthOK: 0, logHealthWarn: 1, logHealthDrift: 2, logHealthError: 3}

func worstLogHealthStatus(statuses []string) string {
	worst := logHealthOK
	for _, status := range statuses {
		if logHealthStatusRank[status] > logHealthStatusRank[worst] {
			worst = status
		}
	}
	return worst
}

func logHealthItem(name, status, detail string) gin.H {
	return gin.H{"name": name, "status": status, "detail": detail}
}

func logHealthLayer(key, name, status, summary string, items []gin.H) gin.H {
	if items == nil {
		items = []gin.H{}
	}
	return gin.H{"key": key, "name": name, "status": status, "summary": summary, "items": items}
}

func logHealthLayerFromItems(key, name string, items []gin.H, emptySummary string) gin.H {
	if len(items) == 0 {
		return logHealthLayer(key, name, logHealthWarn, emptySummary, items)
	}
	abnormal := 0
	statuses := make([]string, 0, len(items))
	for _, item := range items {
		statuses = append(statuses, item["status"].(string))
		if item["status"] != logHealthOK {
			abnormal++
		}
	}
	summary := fmt.Sprintf("%d 项全部一致", len(items))
	if abnormal > 0 {
		summary = fmt.Sprintf("%d/%d 项需要处理", abnormal, len(items))
	}
	return logHealthLayer(key, name, worstLogHealthStatus(statuses), summary, items)
}

// isOpenSearchNotFound：openSearchRequest 的错误形如 "opensearch 404 Not Found: <body>"。
func isOpenSearchNotFound(err error) bool {
	return strings.HasPrefix(err.Error(), "opensearch 404")
}

func (handler *Handler) OpenSearchLogHealth(context *gin.Context) {
	cluster, err := handler.loadOpenSearchCluster(context)
	if err == sql.ErrNoRows {
		response.BusinessError(context, 404, "OpenSearch cluster not found", nil)
		return
	}
	if err != nil {
		response.Error(context, err)
		return
	}
	layers := []gin.H{
		handler.checkLogIndexTemplate(context, cluster),
		handler.checkLogISMPolicies(context, cluster),
		handler.checkLogPipelines(context, cluster),
		handler.checkLogHostConfigs(context),
		handler.checkLogRuntime(context),
		handler.checkLogDataFlow(context, cluster),
	}
	response.Success(context, gin.H{
		"status":     worstLogHealthStatus(layerStatuses(layers)),
		"checked_at": time.Now().UTC().Format(time.RFC3339),
		"layers":     layers,
	})
}

func layerStatuses(layers []gin.H) []string {
	statuses := make([]string, 0, len(layers))
	for _, layer := range layers {
		statuses = append(statuses, layer["status"].(string))
	}
	return statuses
}

func logHealthPrefix(cluster openSearchCluster) string {
	if strings.TrimSpace(cluster.IndexPrefix) == "" {
		return "logs"
	}
	return cluster.IndexPrefix
}

// checkLogIndexTemplate 模板决定字段类型：dynamic 漏成 true 会把 keyword 建成 text，聚合功能全部失效。
func (handler *Handler) checkLogIndexTemplate(context *gin.Context, cluster openSearchCluster) gin.H {
	prefix := logHealthPrefix(cluster)
	name := buildIndexTemplateName(prefix)
	responseBody, err := handler.openSearchRequest(context, cluster, "GET", "/_index_template/"+safeIndexSegment(prefix), nil)
	if err != nil {
		if isOpenSearchNotFound(err) {
			return logHealthLayer("index_template", "索引模板", logHealthDrift, fmt.Sprintf("模板 %s 不存在，新建索引会走动态映射", name), nil)
		}
		return logHealthLayer("index_template", "索引模板", logHealthError, fmt.Sprintf("读取模板失败: %v", err), nil)
	}
	desiredBody := buildIndexTemplateBody(prefix)
	desiredTemplate := desiredBody["template"].(gin.H)["mappings"].(gin.H)
	actualMappings := map[string]any{}
	if templates, ok := responseBody["index_templates"].([]any); ok && len(templates) > 0 {
		if first, ok := templates[0].(map[string]any); ok {
			if indexTemplate, ok := first["index_template"].(map[string]any); ok {
				if template, ok := indexTemplate["template"].(map[string]any); ok {
					mappings, _ := template["mappings"].(map[string]any)
					actualMappings = mappings
				}
			}
		}
	}
	items := []gin.H{}
	// JSON 反序列化后 dynamic 缺省即 true，与 Django get('dynamic', True) 语义一致。
	desiredDynamic, _ := desiredTemplate["dynamic"].(bool)
	actualDynamic := true
	if value, ok := actualMappings["dynamic"].(bool); ok {
		actualDynamic = value
	}
	if actualDynamic == desiredDynamic {
		items = append(items, logHealthItem("dynamic", logHealthOK, fmt.Sprintf("期望 %v，实际 %v", desiredDynamic, actualDynamic)))
	} else {
		items = append(items, logHealthItem("dynamic", logHealthDrift, fmt.Sprintf("期望 %v，实际 %v", desiredDynamic, actualDynamic)))
	}
	actualProperties, _ := actualMappings["properties"].(map[string]any)
	for field, specRaw := range desiredTemplate["properties"].(gin.H) {
		spec := specRaw.(gin.H)
		expectedType := fmt.Sprint(spec["type"])
		current, _ := actualProperties[field].(map[string]any)
		if current == nil {
			items = append(items, logHealthItem(field, logHealthDrift, fmt.Sprintf("缺失，期望 %s", expectedType)))
			continue
		}
		currentType, _ := current["type"].(string)
		if currentType != expectedType {
			items = append(items, logHealthItem(field, logHealthDrift, fmt.Sprintf("类型不符：期望 %s，实际 %s", expectedType, currentType)))
			continue
		}
		items = append(items, logHealthItem(field, logHealthOK, expectedType))
	}
	sort.Slice(items, func(i, j int) bool { return items[i]["name"].(string) < items[j]["name"].(string) })
	return logHealthLayerFromItems("index_template", "索引模板", items, "模板无字段定义")
}

// asStringMap / asAnySlice：gin.H 与 json 反序列化的 map[string]any 是不同动态类型，
// 直接断言会失败，这里统一归一（gin.H 的底层类型就是 map[string]any）。
func asStringMap(value any) map[string]any {
	switch typed := value.(type) {
	case map[string]any:
		return typed
	case gin.H:
		return typed
	default:
		return nil
	}
}

func asAnySlice(value any) []any {
	if typed, ok := value.([]any); ok {
		return typed
	}
	return nil
}

// ismPolicySignature 只提取影响保留行为的字段：集群会补 policy_id/last_updated_time 等元数据，全量比对必然误报。
func ismPolicySignature(policy map[string]any) gin.H {
	signature := gin.H{"rollover_size": nil, "rollover_age": nil, "delete_after": nil, "index_patterns": []string{}}
	hot := map[string]any{}
	for _, stateRaw := range asAnySlice(policy["states"]) {
		state := asStringMap(stateRaw)
		if name, _ := state["name"].(string); name == "hot" {
			hot = state
		}
	}
	for _, actionRaw := range asAnySlice(hot["actions"]) {
		action := asStringMap(actionRaw)
		if rollover := asStringMap(action["rollover"]); rollover != nil {
			signature["rollover_size"] = rollover["min_primary_shard_size"]
			signature["rollover_age"] = rollover["min_index_age"]
			break
		}
	}
	for _, transitionRaw := range asAnySlice(hot["transitions"]) {
		transition := asStringMap(transitionRaw)
		if stateName, _ := transition["state_name"].(string); stateName == "delete" {
			conditions := asStringMap(transition["conditions"])
			signature["delete_after"] = conditions["min_index_age"]
			break
		}
	}
	patterns := []string{}
	templates := asAnySlice(policy["ism_template"])
	if templateDict := asStringMap(policy["ism_template"]); templateDict != nil {
		templates = []any{templateDict}
	}
	for _, templateRaw := range templates {
		template := asStringMap(templateRaw)
		for _, patternRaw := range asAnySlice(template["index_patterns"]) {
			patterns = append(patterns, fmt.Sprint(patternRaw))
		}
	}
	sort.Strings(patterns)
	signature["index_patterns"] = patterns
	return signature
}

// checkLogISMPolicies ISM 缺失时索引既不滚动也不清理，磁盘会被慢慢撑满，属于静默故障。
func (handler *Handler) checkLogISMPolicies(context *gin.Context, cluster openSearchCluster) gin.H {
	prefix := logHealthPrefix(cluster)
	tiers, err := handler.loadEnabledRetentionTiers(context)
	if err != nil {
		return logHealthLayer("ism_policies", "保留策略", logHealthError, "读取保留档位失败: "+err.Error(), nil)
	}
	items := []gin.H{}
	for _, tier := range tiers {
		name := buildISMPolicyName(prefix, tier.Code)
		desired := ismPolicySignature(buildISMPolicyBody(prefix, tier)["policy"].(gin.H))
		remote, err := handler.openSearchRequest(context, cluster, "GET", "/_plugins/_ism/policies/"+name, nil)
		if err != nil {
			if isOpenSearchNotFound(err) {
				items = append(items, logHealthItem(name, logHealthDrift, "策略不存在，索引不会自动滚动与清理"))
			} else {
				items = append(items, logHealthItem(name, logHealthError, truncateOpenSearchError(err)))
			}
			continue
		}
		policy, _ := remote["policy"].(map[string]any)
		if policy == nil {
			policy = map[string]any{}
		}
		actual := ismPolicySignature(policy)
		differing := differingSignatureKeys(actual, desired)
		if len(differing) == 0 {
			items = append(items, logHealthItem(name, logHealthOK, fmt.Sprintf("保留 %d 天", tier.RetentionDays)))
			continue
		}
		items = append(items, logHealthItem(name, logHealthDrift, "与档位配置不一致: "+strings.Join(differing, ", ")))
	}
	return logHealthLayerFromItems("ism_policies", "保留策略", items, "没有启用中的保留档位")
}

func differingSignatureKeys(actual, desired gin.H) []string {
	differing := make([]string, 0, 4)
	for _, key := range []string{"rollover_size", "rollover_age", "delete_after", "index_patterns"} {
		if !sameSignatureValue(actual[key], desired[key]) {
			differing = append(differing, key)
		}
	}
	return differing
}

func sameSignatureValue(actual, desired any) bool {
	switch left := actual.(type) {
	case string:
		right, ok := desired.(string)
		return ok && left == right
	case float64:
		right, ok := desired.(float64)
		return ok && left == right
	case bool:
		right, ok := desired.(bool)
		return ok && left == right
	case nil:
		return desired == nil
	default:
		leftJSON, _ := json.Marshal(actual)
		rightJSON, _ := json.Marshal(desired)
		return string(leftJSON) == string(rightJSON)
	}
}

// checkLogPipelines pipeline 没发布时日志照样写入，只是不被解析，界面上看不出任何异常。
func (handler *Handler) checkLogPipelines(context *gin.Context, cluster openSearchCluster) gin.H {
	type processingRuleRow struct {
		Name         string
		PipelineBody json.RawMessage
	}
	rows, err := handler.db.QueryContext(context, `SELECT name,pipeline_body FROM monitor_log_processing_rule WHERE cluster_id=? ORDER BY name`, cluster.ID)
	if err != nil {
		return logHealthLayer("pipelines", "解析规则", logHealthError, "读取解析规则失败: "+err.Error(), nil)
	}
	rules := make([]processingRuleRow, 0, 8)
	for rows.Next() {
		var rule processingRuleRow
		if err := rows.Scan(&rule.Name, &rule.PipelineBody); err != nil {
			rows.Close()
			return logHealthLayer("pipelines", "解析规则", logHealthError, "读取解析规则失败: "+err.Error(), nil)
		}
		rules = append(rules, rule)
	}
	rows.Close()

	items := []gin.H{}
	for _, rule := range rules {
		remote, err := handler.openSearchRequest(context, cluster, "GET", "/_ingest/pipeline/"+rule.Name, nil)
		if err != nil {
			if isOpenSearchNotFound(err) {
				items = append(items, logHealthItem(rule.Name, logHealthDrift, "集群上不存在该 pipeline，日志不会被解析"))
			} else {
				items = append(items, logHealthItem(rule.Name, logHealthError, truncateOpenSearchError(err)))
			}
			continue
		}
		body, _ := remote[rule.Name].(map[string]any)
		if body == nil {
			items = append(items, logHealthItem(rule.Name, logHealthDrift, "集群上不存在该 pipeline，日志不会被解析"))
		} else if pipelineSignature(body) != pipelineSignature(json.RawMessage(rule.PipelineBody)) {
			items = append(items, logHealthItem(rule.Name, logHealthDrift, "集群上的 pipeline 与页面配置不一致，需重新发布"))
		} else {
			var decoded map[string]any
			_ = json.Unmarshal(rule.PipelineBody, &decoded)
			processors, _ := decoded["processors"].([]any)
			items = append(items, logHealthItem(rule.Name, logHealthOK, fmt.Sprintf("%d 个处理器", len(processors))))
		}
	}
	return logHealthLayerFromItems("pipelines", "解析规则", items, "尚未配置解析规则")
}

// checkLogHostConfigs 比对数据库记录，暴露未安装/未下发配置的主机。
// 注意：Go 侧 Fluent Bit 配置由 agent 渲染（configure_fluent_bit_opensearch），
// 后端没有期望指纹生成器，因此这里只判断「已下发与否」，不比对内容一致性；
// 主机被人手工改过的情况由 agent 侧上报与数据流层兜底。
func (handler *Handler) checkLogHostConfigs(context *gin.Context) gin.H {
	type targetRow struct {
		ID               int64
		HostIP           string
		AgentInstalled   bool
		ConfigFingerprint string
	}
	rows, err := handler.db.QueryContext(context, `SELECT l.id,COALESCE(h.ip,''),l.agent_installed,COALESCE(l.config_fingerprint,'') FROM monitor_log_collection_target l JOIN assets_host h ON h.id=l.host_id WHERE l.managed_enabled=TRUE ORDER BY l.id`)
	if err != nil {
		return logHealthLayer("host_configs", "主机配置", logHealthError, "读取采集目标失败: "+err.Error(), nil)
	}
	targets := make([]targetRow, 0, 16)
	for rows.Next() {
		var target targetRow
		if err := rows.Scan(&target.ID, &target.HostIP, &target.AgentInstalled, &target.ConfigFingerprint); err != nil {
			rows.Close()
			return logHealthLayer("host_configs", "主机配置", logHealthError, "读取采集目标失败: "+err.Error(), nil)
		}
		targets = append(targets, target)
	}
	rows.Close()

	items := []gin.H{}
	for _, target := range targets {
		label := target.HostIP
		if label == "" {
			label = fmt.Sprintf("host-%d", target.ID)
		}
		if !target.AgentInstalled {
			items = append(items, logHealthItem(label, logHealthWarn, "Fluent Bit 未安装"))
			continue
		}
		if strings.TrimSpace(target.ConfigFingerprint) == "" {
			items = append(items, logHealthItem(label, logHealthDrift, "从未下发过采集配置"))
			continue
		}
		items = append(items, logHealthItem(label, logHealthOK, "已下发采集配置"))
	}
	return logHealthLayerFromItems("host_configs", "主机配置", items, "没有纳管中的采集目标")
}

// checkLogRuntime 运行状态取自数据库缓存，反映最近一次探测结果，不是实时探活。
func (handler *Handler) checkLogRuntime(context *gin.Context) gin.H {
	type targetRow struct {
		ID         int64
		HostIP     string
		Runtime    string
		LastError  string
	}
	rows, err := handler.db.QueryContext(context, `SELECT l.id,COALESCE(h.ip,''),COALESCE(l.runtime_status,''),COALESCE(l.last_error,'') FROM monitor_log_collection_target l JOIN assets_host h ON h.id=l.host_id WHERE l.managed_enabled=TRUE AND l.agent_installed=TRUE ORDER BY l.id`)
	if err != nil {
		return logHealthLayer("runtime", "采集进程", logHealthError, "读取采集目标失败: "+err.Error(), nil)
	}
	targets := make([]targetRow, 0, 16)
	for rows.Next() {
		var target targetRow
		if err := rows.Scan(&target.ID, &target.HostIP, &target.Runtime, &target.LastError); err != nil {
			rows.Close()
			return logHealthLayer("runtime", "采集进程", logHealthError, "读取采集目标失败: "+err.Error(), nil)
		}
		targets = append(targets, target)
	}
	rows.Close()

	items := []gin.H{}
	for _, target := range targets {
		label := target.HostIP
		if label == "" {
			label = fmt.Sprintf("host-%d", target.ID)
		}
		switch target.Runtime {
		case "running":
			items = append(items, logHealthItem(label, logHealthOK, "运行中"))
		case "error":
			message := target.LastError
			if strings.TrimSpace(message) == "" {
				message = "异常"
			}
			if len(message) > 200 {
				message = message[:200]
			}
			items = append(items, logHealthItem(label, logHealthError, message))
		case "stopped":
			items = append(items, logHealthItem(label, logHealthDrift, "已停止"))
		default:
			items = append(items, logHealthItem(label, logHealthWarn, "状态未知，需刷新"))
		}
	}
	return logHealthLayerFromItems("runtime", "采集进程", items, "没有已安装 Fluent Bit 的主机")
}

// checkLogDataFlow 前面几层全绿也可能没数据，这一层是唯一能证明链路真正通了的证据。
func (handler *Handler) checkLogDataFlow(context *gin.Context, cluster openSearchCluster) gin.H {
	prefix := logHealthPrefix(cluster)
	body := gin.H{
		"size":  0,
		"query": gin.H{"range": gin.H{"@timestamp": gin.H{"gte": fmt.Sprintf("now-%dm", logHealthDataFlowWindowMinutes)}}},
		"aggs":  gin.H{"by_service": gin.H{"terms": gin.H{"field": "service", "size": 50}}},
	}
	result, err := handler.openSearchRequest(context, cluster, "POST", "/"+prefix+"-*/_search", body)
	if err != nil {
		return logHealthLayer("data_flow", "数据写入", logHealthError, fmt.Sprintf("查询失败: %v", err), nil)
	}
	total := 0.0
	if hits, ok := result["hits"].(map[string]any); ok {
		if totalRaw, ok := hits["total"].(map[string]any); ok {
			total, _ = totalRaw["value"].(float64)
		}
	}
	buckets := []any{}
	if aggregations, ok := result["aggregations"].(map[string]any); ok {
		if byService, ok := aggregations["by_service"].(map[string]any); ok {
			buckets, _ = byService["buckets"].([]any)
		}
	}
	items := []gin.H{}
	for _, bucketRaw := range buckets {
		bucket, _ := bucketRaw.(map[string]any)
		name, _ := bucket["key"].(string)
		count, _ := bucket["doc_count"].(float64)
		items = append(items, logHealthItem(name, logHealthOK, fmt.Sprintf("%.0f 条", count)))
	}
	if total == 0 {
		return logHealthLayer("data_flow", "数据写入", logHealthWarn, fmt.Sprintf("最近 %d 分钟没有新日志写入", logHealthDataFlowWindowMinutes), items)
	}
	return logHealthLayer("data_flow", "数据写入", logHealthOK, fmt.Sprintf("最近 %d 分钟写入 %.0f 条，覆盖 %d 个服务", logHealthDataFlowWindowMinutes, total, len(items)), items)
}
